Make stacks updateable

fghaas opened the pull request:
When we configure servers (via `OS::Nova::Server`) for boot-from-volume, and we actually want to use a single persistent volume to boot from, then this creates a catch-22 when we update the server resource in such a way that Heat detects that the server requires replacement: Heat then tries to spin up a new server before removing the old one, but at that stage it obviously cannot attach the boot volume.

One way in which we can circumvent this problem is to use BFV, but to not actually boot from a single boot volume, but rather to recreate the boot volume each time the server is recreated.

Thus:

* Drop the `my_dockerhost_boot_volume` resource.
* Instead, create the boot volume directly from the `block_device_mapping_v2` map in the `OS::Nova::Server` resource.

We still need a _separate_ volume — not for booting the host, but for Docker to actually manage containers on. For that, however, we can't use `block_device_mapping_v2`; we must use an `OS::Cinder::Volume` resource instead.

Finally, while we're at it, we can add a `WaitCondition` resource to indicate when the entire configuration of the stack (including that of the Docker host) is complete.

Commit:
Make stacks updateable

When we configure servers (via OS::Nova::Server) for boot-from-volume,
and we actually want to use a single persistent volume to boot from,
then this creates a catch-22 when we update the server resource in
such a way that Heat detects that the server requires replacement:
Heat then tries to spin up a new server before removing the old one,
but at that stage it obviously cannot attach the boot volume.

One way in which we can circumvent this problem is to use BFV, but to
not actually boot from a single boot volume, but rather to recreate
the boot volume each time the server is recreated.

Thus:

* Drop the my_dockerhost_boot_volume resource.
* Instead, create the boot volume directly from the
  block_device_mapping_v2 map in the OS::Nova::Server resource.

colder-is-better commented on hot/dockerhost-stack-1.yaml:
```diff
@@ -38,15 +38,10 @@ resources:
         - device_name: vda
-          volume_id: { get_resource: my_dockerhost_boot_volume }
+          image: { get_param: image }
+          volume_size: { get_param: boot_volume_size }
```
```suggestion
          volume_size: { get_param: boot_volume_size }
          delete_on_termination: true
```

colder-is-better commented on hot/dockerhost-stack-2.yaml:
```diff
@@ -38,15 +38,10 @@ resources:
         - device_name: vda
-          volume_id: { get_resource: my_dockerhost_boot_volume }
+          image: { get_param: image }
+          volume_size: { get_param: boot_volume_size }
```
```suggestion
          volume_size: { get_param: boot_volume_size }
          delete_on_termination: true
```

colder-is-better commented on hot/dockerhost-stack-3.yaml:
```diff
@@ -38,17 +38,12 @@ resources:
         - device_name: vda
-          volume_id: { get_resource: my_dockerhost_boot_volume }
+          image: { get_param: image }
+          volume_size: { get_param: boot_volume_size }
```
```suggestion
          volume_size: { get_param: boot_volume_size }
          delete_on_termination: true
```

colder-is-better commented on hot/dockerhost-stack.yaml:
```diff
@@ -38,17 +38,12 @@ resources:
         - device_name: vda
-          volume_id: { get_resource: my_dockerhost_boot_volume }
+          image: { get_param: image }
+          volume_size: { get_param: boot_volume_size }
```
```suggestion
          volume_size: { get_param: boot_volume_size }
          delete_on_termination: true
```

colder-is-better reviewed (commented):
The proposed workaround works fine -- I went through the whole process in the tutorial and tested all templates. I am adding, though, the `delete_on_termination: true` assignment in each `block_device_mapping_v2` list, otherwise, it would be kind of weird to explain the dangling no-name volume when I talk about stack deletion.

Commit:
Configure boot volumes to auto-delete on server termination

If we leave delete_on_termination at its default (false), then this
creates dangling orphan volumes whenever a stack update results in the
replacement of a server, and in addition all of the stack's volumes
survive the deletion of the whole stack. Thus, flip it to true.

fghaas commented:
I've pushed bc29cdf56408b8be7ad1c8a3393563a4cb9e3d54 to add `delete_on_termination: true`.

colder-is-better reviewed (approved):
Very nice, thanks for those additions.

Commit:
Add a persistent volume for use by Docker

In order to ensure that Docker has something to work with that
persists across server replacements, add a (non-boot) persistent
volume via an OS::Cinder::Volume resource, and manage it via
OS::Cinder::VolumeAttachment.

Then, use cloud-init to manage this volume as a btrfs filesystem, and
mount it to /var/lib/docker.

Commit:
Add a WaitCondition

Without a WaitCondition resource, Heat considers the creation of the
stack (and any subsequent stack updates) complete as soon as the
OS::Nova::Server instance has booted. This is not usually helpful;
instead we want things to be considered "done" when the stack has been
fully configured.

The most straightforward way to do that in a Heat context is to add a
WaitCondition, and invoke it at the end of the cloud-init sequence via
runcmd.
